### ISSUE TITLE
Add support for getting the default ScrollController from context using ScrollController.of()

### ISSUE THREAD

**Issue by amdshrif** [flutter/flutter#118620]
## Use case

Consider a scenario where we have a nested CustomScrollView inside a NestedScrollView. The CustomScrollView has a child that needs to know the current scroll position of the CustomScrollView to perform some action. Without the proposed feature, the developer would have to manually pass the ScrollController of the CustomScrollView through the widget tree to the child that needs it. This can lead to a lot of boilerplate code, especially when dealing with complex nested scroll views.

With the proposed feature, the developer can simply use ScrollController.of(context) in the child that needs to access the scroll position. This eliminates the need for manually passing the ScrollController through the widget tree and makes it much easier to access the scroll position of a nested CustomScrollView.

This is a common use case, when using NestedScrollView which is a built-in scrollable that can contain multiple scrolling widgets, and the CustomScrollView which allows developers to create custom scroll effects.

This use case will also help to improve the code readability and structure, and would make it much more straightforward to access the ScrollController in a complex nested scroll view.

### Introduction:

It would be useful to have a way to easily access the default ScrollController of a scrollable widget from the context, similar to how we can access the Theme using Theme.of(context). This would allow developers to easily access the ScrollController of a ScrollView, for example, without having to manually pass it down through the widget tree.

### Proposed solution:

Introduce a new method, ScrollController.of(context), that takes a BuildContext as an argument and returns the default ScrollController of the nearest ancestor Scrollable widget. If no such widget is found, the method should return null.

### Benefit:

This feature will be useful for developers who want to access the ScrollController of a ScrollView without having to manually pass it down through the widget tree. This can also help reduce the amount of boilerplate code needed to manage the ScrollController of a ScrollView.

### Impact:

This feature will have a low impact on the existing codebase and will be easy to implement. It will also improve developer productivity by reducing the amount of boilerplate code needed to manage the ScrollController of a ScrollView.

### Example Code:
In this example, MyChild is a widget that needs to know the current scroll position of the nearest ancestor Scrollable widget, which in this case is the CustomScrollView. Without the proposed feature, the developer would have to manually pass the ScrollController of the CustomScrollView through the widget tree to MyChild. But with the proposed feature, the developer can simply use ScrollController.of(context) in MyChild, eliminating the need for manually passing the ScrollController through the widget tree.

```dart
import 'package:flutter/material.dart';

class MyChild extends StatefulWidget {
  MyChild({Key? key}) : super(key: key);

  @override
  State<MyChild> createState() => _MyChildState();
}

class _MyChildState extends State<MyChild> {
  void _onScrolled() {
    setState(() {
      print('Scrolling');
    });
  }

  @override
  Widget build(BuildContext context) {
    final ScrollController controller = ScrollController.of(context);
    final double currentPosition = controller.position.pixels;
    controller.addListener(_onScrolled);
    // return Text("Current scroll position: $currentPosition");
    try {
      return SliverAppBar(
        pinned: true,
        title: Text("Current position: $currentPosition"),
      );
    } finally {
      controller.removeListener(_onScrolled);
    }
  }
}

class MyParent extends StatelessWidget {
  final ScrollController _controller = ScrollController();

  @override
  Widget build(BuildContext context) {
    return NestedScrollView(
      floatHeaderSlivers: true,
      headerSliverBuilder: (BuildContext context, bool innerBoxIsScrolled) {
        return [
          const SliverAppBar(
            title: Text("MyParent"),
            floating: true,
            // expandedHeight: 150,
          ),
        ];
      },
      body: CustomScrollView(
        controller: _controller,
        slivers: [
          MyChild(),
          SliverList.builder(
            itemCount: 9,
            itemBuilder: (context, index) {
              return Container(height: 200, color: Colors.blue[100 * index]);
            },
          ),
        ],
      ),
    );
  }
}

```

Here's an example of how a developer would have to manually pass the ScrollController of the CustomScrollView to the child:

```dart
void main() {
  runApp(MaterialApp(home: MyParent()));
}

class MyChild extends StatefulWidget {
  final ScrollController controller;

  const MyChild({Key? key, required this.controller}) : super(key: key);

  @override
  State<MyChild> createState() => _MyChildState();
}

class _MyChildState extends State<MyChild> {
  void _onScrolled() {
    setState(() {
      print('Scrolling');
    });
  }

  @override
  void initState() {
    super.initState();
    widget.controller.addListener(_onScrolled);
  }

  @override
  void dispose() {
    widget.controller.removeListener(_onScrolled);
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final double currentPosition = widget.controller.position.pixels;
    return SliverAppBar(
      pinned: true,
      title: Text("Current position: $currentPosition"),
    );
  }
}

class MyParent extends StatelessWidget {
  final ScrollController _controller = ScrollController();

  @override
  Widget build(BuildContext context) {
    return NestedScrollView(
      floatHeaderSlivers: true,
      headerSliverBuilder: (BuildContext context, bool innerBoxIsScrolled) {
        return [
          const SliverAppBar(
            title: Text("MyParent"),
            floating: true,
            // expandedHeight: 150,
          ),
        ];
      },
      body: CustomScrollView(
        controller: _controller,
        slivers: [
          MyChild(controller: _controller),
          SliverList.builder(
            itemCount: 9,
            itemBuilder: (context, index) {
              return Container(height: 200, color: Colors.blue[100 * index]);
            },
          ),
        ],
      ),
    );
  }
}

void main() {
  runApp(MaterialApp(home: MyParent()));
}
```

**Comment by amdshrif** [comment]
@darshankawarI am willing to handle this issue, but I would appreciate guidance and assistance as it would be my first contribution. You can assign it to me.

**Comment by clragon** [comment]
Hi,

Can you explain how this is different from [PrimaryScrollController](https://api.flutter.dev/flutter/widgets/PrimaryScrollController-class.html)?

`NestedScrollView` does already provide a `PrimaryScrollController` for its children, which you can then access with `PrimayScrollController.of(context)` and should already be used by default in any `ScrollView` that is set as `primary: true`.

**Comment by amdshrif** [comment]
@clragon I have tried using `PrimaryScrollController` but I am getting this error:

```
======== Exception caught by widgets library =======================================================
The following assertion was thrown building MyChild(dirty, dependencies: [PrimaryScrollController], state: _MyChildState#4cb75):
PrimaryScrollController.of() was called with a context that does not contain a PrimaryScrollController widget.

No PrimaryScrollController widget ancestor could be found starting from the context that was passed to PrimaryScrollController.of(). This can happen because you are using a widget that looks for a PrimaryScrollController ancestor, but no such ancestor exists.
The context used was:
  MyChild(dirty, dependencies: [PrimaryScrollController], state: _MyChildState#4cb75)
The relevant error-causing widget was: 
```
Here is the code:

```Dart
import 'package:flutter/material.dart';

void main() {
  runApp(MaterialApp(home: MyParent()));
}

class MyChild extends StatefulWidget {
  MyChild({Key? key}) : super(key: key);

  @override
  State<MyChild> createState() => _MyChildState();
}

class _MyChildState extends State<MyChild> {

  void _onScrolled() {
    setState(() {
      print('Scrolling');
    });
  }

  @override
  Widget build(BuildContext context) {
    final ScrollController controller = PrimaryScrollController.of(context);
    final double currentPosition = controller.position.pixels;
    controller.addListener(_onScrolled);
    try {
      return SliverAppBar(
        pinned: true,
        title: Text("Current position: $currentPosition"),
      );
    } finally {
      controller.removeListener(_onScrolled);
    }
  }
}

class MyParent extends StatelessWidget {
  final ScrollController _controller = ScrollController();

  @override
  Widget build(BuildContext context) {
    return NestedScrollView(
      floatHeaderSlivers: true,
      headerSliverBuilder: (BuildContext context, bool innerBoxIsScrolled) {
        return [
          const SliverAppBar(
            title: Text("MyParent"),
            floating: true,
          ),
        ];
      },
      body: PrimaryScrollController(
        controller: _controller,
        child: CustomScrollView(
          primary: true,
          slivers: [
            MyChild(),
            SliverList.builder(
              itemCount: 9,
              itemBuilder: (context, index) {
                return Container(height: 200, color: Colors.blue[100 * index]);
              },
            ),
          ],
        ),
      ),
    );
  }
}

```

As you can see the error says that the `PrimaryScrollController.of() `was called with a context that does not contain a PrimaryScrollController widget but that looks not true to me, the `MyChild` is within the subtree of  `PrimaryScrollController` widget as you can see in the example code.

**Comment by clragon** [comment]
you are incorrectly using `PrimaryScrollController`.
`CustomScrollView` will remove the `PrimaryScrollController` for its descendants once it has consumed it. Its also not necessary to provide a `PrimaryScrollController` as `NestedScrollView` already does so, as I have mentioned above.
It is also not necessary to specify the controller of `NestedScrollView`:

```dart
    return NestedScrollView(
      floatHeaderSlivers: true,
      headerSliverBuilder: (BuildContext context, bool innerBoxIsScrolled) {
        return [
          const SliverAppBar(
            title: Text("MyParent"),
            floating: true,
          ),
        ];
      },
      body: CustomScrollView(
          primary: true,
          slivers: [
            MyChild(),
            SliverList.builder(
              itemCount: 9,
              itemBuilder: (context, index) {
                return Container(height: 200, color: Colors.blue[100 * index]);
              },
            ),
          ],
        ),
      ),
    );
```

The flutter/flutter github repository is for issues with the framework. Its not intended to be a place for helping with how to use its components.
If you need help fixing your code please instead visit any of the resources listed at https://flutter.dev/community.

**Comment by amdshrif** [comment]
@clragon in https://github.com/flutter/flutter/issues/118620#issuecomment-1399503460 I just shared with you the reason why I did not use `PrimaryScrollController` and your feedback above exactly answers your question here https://github.com/flutter/flutter/issues/118620#issuecomment-1396925978.

And for your suggestion here https://github.com/flutter/flutter/issues/118620#issuecomment-1399557695, you are welcome to try it yourself and share the result as I have tired that using master and here is what I get:

```
======== Exception caught by widgets library =======================================================
The following assertion was thrown building MyChild(dirty, dependencies: [PrimaryScrollController], state: _MyChildState#103bf):
PrimaryScrollController.of() was called with a context that does not contain a PrimaryScrollController widget.

No PrimaryScrollController widget ancestor could be found starting from the context that was passed to PrimaryScrollController.of(). This can happen because you are using a widget that looks for a PrimaryScrollController ancestor, but no such ancestor exists.
The context used was:
  MyChild(dirty, dependencies: [PrimaryScrollController], state: _MyChildState#103bf)
The relevant error-causing widget was: 
``` 

To sum up, for these reasons I suggested to add support for getting the default ScrollController from context using ScrollController.of() so we can have more flexibility.

P.S. I am aware of the purpose of this repository and I am familiar with where to find help on how to use Flutter. The issue title is to include support for a specific scenario mentioned in the proposal, and I am confident that I am in the right place. Let's stay on topic and not deviate from the scenario discussed in the proposal.

**Comment by clragon** [comment]
I see. Of course, because `CustomScrollView` swallows the `PrimaryScrollController`, you will not be able to access it in children of `CustomScrollView`. Thats why you would like to have a generalized concept of accessing the nearest `ScrollView`'s controller.

**Comment by goderbauer** [comment]
cc @Piinks

**Comment by amdshrif** [comment]
> I see. Of course, because `CustomScrollView` swallows the `PrimaryScrollController`, you will not be able to access it in children of `CustomScrollView`. Thats why you would like to have a generalized concept of accessing the nearest `ScrollView`'s controller.
> 
> Wrapping all children of `CustomScrollView` in a `PrimaryScrollController` would be impractical and against the idea of it. So thats why a new way of getting the surrounding `ScrollController` is desirable.

You got it.

**Comment by snipd-mikel** [comment]
Maybe not what you need, but for some use cases where you need the ScrollController (for example, to scroll to a given offset), you can also use `Scrollable.of(context)` or `Scrollable.maybeOf(context)`.

You can then use it's position to jump to an offset or get other information about the scroll position:

```dart
 Scrollable.maybeOf(context)?.position.animateTo(
                0,
                duration: const Duration(milliseconds: 500),
                curve: Curves.easeInOutQuad,
              );

```
